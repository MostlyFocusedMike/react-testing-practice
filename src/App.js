import React, { useState, useEffect } from 'react';
import './App.css';

const App = () => {
    const [user, setUser] = useState(null)
    const [userId, setUserId] = useState('');

    const getUser = async (id) => {
        const user = await fetch(`https://jsonplaceholder.typicode.com/users/${id}`).then(response => response.json())
        setUser(user);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if (!userId) return alert('Enter an id!');
        getUser(userId);
    }

    const handleChange = (e) => {
        const id = parseInt(e.target.value);
        if (isNaN(id)) setUserId('');
        if (id < 11 && id > 0) setUserId(id);
    }

    useEffect(() => {
        getUser(1);
    }, [])

    const userCard = () => {
        return (
            <>
                <h1>{user.name}</h1>
                <h2>{user.email}</h2>
                <p>{user.username}</p>
            </>
        );
    }
    return (
        <div className="App">
            <h1>React Testing Example</h1>
            {
                user
                    ? userCard()
                    : <h1>Enter an id</h1>
            }
            <form onSubmit={handleSubmit} >
                <label htmlFor='user-id'>User Id</label>
                <input id='user-id' value={userId} onChange={handleChange}></input>
                <button>Submit</button>
            </form>
        </div>
    );
}

export default App;
