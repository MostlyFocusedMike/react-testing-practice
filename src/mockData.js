export const mockUser1 = {
    email: "Sincere@april.biz",
    id: 1,
    name: "Leanne Graham",
    phone: "1-770-736-8031",
    username: "Bret",
    website: "hildegard.org",
}

export const mockUser2 = {
    email: "Shanna@melissa.tv",
    id: 2,
    name: "Ervin Howell",
    phone: "010-692-6593 x09125",
    username: "Antonette",
    website: "anastasia.net",
}
