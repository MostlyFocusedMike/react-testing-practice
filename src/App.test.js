import React from 'react';
import { enableFetchMocks } from 'jest-fetch-mock'
import { render, screen, fireEvent, waitForElement } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';
import { mockUser1, mockUser2 } from './mockData';

enableFetchMocks(); // mocks the fetch method for the test

describe('Basic Tests', () => {
    beforeAll(() => {
        // Five months ago jest-fetch-mock 3.0.0 FINALLY added the ability to mock by route,
        // not request order. I *highly* recommend mocking by route, since tests will
        // only succeed if the actual fetched url is correct. Fetches that are mocked
        // by order ignore the fetch destination so some bugs can sneak through.

        // use regex to grab routes dynamically
        const userRoutesUrl = /^https:\/\/jsonplaceholder.typicode.com\/users\/[\d]$/;

        const mockUsersRoutes = async (req) => {
            if (req.url.endsWith(mockUser1.id)) return JSON.stringify(mockUser1);
            if (req.url.endsWith(mockUser2.id)) return JSON.stringify(mockUser2);
            return {
                status: 404,
                body: "Not Found"
            }
        };

        fetch.mockIf(userRoutesUrl, mockUsersRoutes);
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    const setup = async () => {
        const { container } = render(<App />);

        await waitForElement(() => screen.getByText(mockUser1.name))
        const submitButton = screen.getByText(/Submit/)
        const idInput = screen.getByLabelText(/User Id/)
        return {
            container,
            submitButton,
            idInput,
        };
    };

    // afterEach(() => {
    //     cleanup() // this step is no longer needed
    // });

    it('renders at all', async () => {
        await setup();
        const mainHeader = screen.getByText(/React Testing Example/i);
        expect(mainHeader).toBeInTheDocument();
    });

    it('Grabs first user', async () => {
        await setup();
        expect(screen.getByText(mockUser1.name)).toBeInTheDocument();
    });

    it('Grabs first user but this time uses a snapshot', async () => {
        const { container } = await setup();
        expect(container.firstChild).toMatchSnapshot();
    });

    it('Successfully submits form to get second user', async () => {
        const { submitButton, idInput } = await setup();

        fireEvent.change(idInput, {
            target: { value: 2 },
        });
        fireEvent.click(submitButton);

        const newname = await screen.findByText(mockUser2.name);
        expect(newname).toBeInTheDocument();
        const initialName = screen.queryByText(mockUser1.name);

        expect(initialName).not.toBeInTheDocument();
    });

    it('Throws an alert when trying to use an empty id', async () => {
        const { submitButton } = await setup();
        // how to mock globals
        global.alert = jest.fn();

        fireEvent.click(submitButton);
        expect(global.alert).toHaveBeenCalledTimes(1);
        expect(global.alert).toHaveBeenCalledWith('Enter an id!');
    });

    describe('Controlled input does not allow values except integers 1-10', () => {
        const initialValue = '';
        const tooLowNum = '0';
        const minNum = '1';
        const maxNum = '10';
        const tooHighNum = '11';

        // hmmmmmm, these tests seem awfully similar. Should we combine them or loop?
        // Not necessarily. While some devs certainly would, tests should be readable and atomic
        // (that means non reliant on any other tests being run). So looping or combining
        // might not help those two goals. But! Others may disagree, and that's ok, just be
        // consistent with your design choices and you'll be alright
        it('accepts 1 as lowest id', async () => {
            const { idInput } = await setup();

            fireEvent.change(idInput, { target: { value: minNum }});
            expect(idInput.value).toBe(minNum);
        });

        it('accepts 10 as the maximum', async () => {
            const { idInput } = await setup();
            // this is the userEvent way to do it
            await userEvent.type(idInput, maxNum);
            expect(idInput.value).toBe(maxNum);
        });

        it('rejects lower than 1 and more than 10', async () => {
            const { idInput } = await setup();

            fireEvent.change(idInput, { target: { value: tooLowNum }});
            expect(idInput.value).toBe(initialValue);

            fireEvent.change(idInput, { target: { value: tooHighNum }});
            expect(idInput.value).toBe(initialValue);
        });

        it('rejects strings', async () => {
            const { idInput } = await setup();

            fireEvent.change(idInput, { target: { value: 'a' }});
            expect(idInput.value).toBe(initialValue);
            fireEvent.change(idInput, { target: { value: 'a?' }});
            expect(idInput.value).toBe(initialValue);
            fireEvent.change(idInput, { target: { value: "look it's whatever, strings fail" }});
            expect(idInput.value).toBe(initialValue);
        });
    });
});
