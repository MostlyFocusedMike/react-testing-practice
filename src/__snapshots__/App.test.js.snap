// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Tests Grabs first user but this time uses a snapshot 1`] = `
<div
  class="App"
>
  <h1>
    React Testing Example
  </h1>
  <h1>
    Leanne Graham
  </h1>
  <h2>
    Sincere@april.biz
  </h2>
  <p>
    Bret
  </p>
  <form>
    <label
      for="user-id"
    >
      User Id
    </label>
    <input
      id="user-id"
      value=""
    />
    <button>
      Submit
    </button>
  </form>
</div>
`;
